test/test_faiss_and_milvus.py

```diff
@@ -205,3 +204,50 @@ def test_faiss_passing_index_from_outside(tmp_path):
         assert 0 <= int(doc.meta["vector_id"]) <= 7
 
 
+def test_faiss_cosine_similarity(tmp_path):
```

Comment:
Could we also add a test to ensure that this implementation of cosine similarity in FAISS also gives the same results as cosine in another document store (e.g. Elasticsearch)?

Reply:
I'm not so sure about that. My impression was that the more content dependent checks are situated in the benchmarks directory. Also, I have some other doubts:

Mathematically, dot product on normalized vectors equals cosine similarity, so we would really be testing the faiss implementation of `normalize_L2` and their dot product search method here. I think testing external dependencies is Ok, it's just not something I usually do in my projects, so I just want to make sure that this is your intention. Also, does it have to be compared to another document store? I just wouldn't expect an elastic search specific test in the faiss test file, it also couples running the tests in a way, that the faiss tests can not run without elastic search. If you want to test for cosine similarity wouldn't it be enough to test against another well tested cosine similarity function such as https://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise.cosine_similarity.html ?

Reply:
Yeah very valid points. The intention is just to have a sanity check that this FAISS cosine similarity method is generating the right numbers. Could we have a check against the sklearn cosine similarity fn then?

Reply:
Why do we even need sklearn? IMO a static example with a known document-query-pair where we check the FAISS cosine similarity result against the static, expected value for the pair should be sufficient. Do I miss anything?

Reply:
It's just a matter of how much we want to test here. But my suggestion for this test is more a nice to have than a crucial feature and comes from the benchmarking perspective. If this introduces an extra dependency then I can see that it would be better to do without it. Happy to go with your suggestion

Reply:
Just in my view tests is not a place to comparative things, but yes it is nice addition things for benchmarking.

Reply:
@mathislucka I think we are very close to the finish line. Do you want to push for the last metres here and add this remaining test to compare if the returned cosine similarity from FAISS equals the expected one (can be a static, known float IMO)? Alternatively, if we should take over, just let us know :)

Reply:
Sorry, I was a bit busy during the last weeks. I will try to implement the test this week. I'll get back to you if something doesn't work out as planned.

Reply:
@tholor I've added a test case. Hope it passes and this is what you had in mind.

Reply:
fixed the typo -.-

Reply:
The test gives me the wrong result. I've digged into the code a little and I think this line is causing the issue:
`            doc.score = float(expit(np.asarray(raw_score / 100)))` (https://github.com/deepset-ai/haystack/blob/3deff26b6057cb9213cb1707cc04acc9849d59d9/haystack/document_store/faiss.py#L429). If I put the expected cosine similarity into this function, I get the same result as in the test (~0.5). Why do you need this line? Would it be enough to not do the transformation when similarity == 'cosine'?

Reply:
Good catch! This normalization doesn't make sense here anymore. We want to make sure that `doc.score` is always in range [0,1] regardless of the docstore or metric being used under the hood. This standardization helps, for example, when you want to show the "relevance" of a search result in a UI to the end-user. This is easier if you can say this score is always between 0-100%. 

So I'd say we normalize in the case of cosine similarity by `score = (sim + 1) / 2` to avoid negative values [-1, 0].

Reply:
Alright, got it.

One thing that worries me a little bit with your normalization method is that it somewhat inflates lower scores. Say we have a cosine similarity of 0.6 this would yield a normalized score of `(0.6 + 1) / 2 = 0.8` meaning the score is inflated by 0.2. In comparison, a score of 0.8 would yield a normalized score of `(0.8 + 1) / 2 = 0.9` which only corresponds to a 0.1 increase. I think this could lead to a false sense of similarity when presenting these normalized scores to the end-user. What do you think?

Reply:
I see your point. Do you have a suggestion for a better conversion function? IMO it's a tradeoff between simplicity of the conversion function, comparability to scores from other metrics/retrievers and balanced upper range for the end-user. One advantage of the proposed linear transformation function would be that the distribution of values remains similar, i.e if the original cosine similarity values follow a normal distribution we can also expect a normal distribution after the transformation. 
@julian-risch any opinion?

Reply:
I would also go for `score = (sim + 1) / 2` for simplicity reasons. Applications built on top of haystack could still apply their on transformation function on top of our scores.

Reply:
@tholor I've added the normalization and updated the tests.